src/liballoc/collections/vec_deque.rs

```diff
@@ -855,8 +856,31 @@ impl<T> VecDeque<T> {
     /// ```
     #[stable(feature = "deque_extras", since = "1.16.0")]
     pub fn truncate(&mut self, len: usize) {
-        for _ in len..self.len() {
-            self.pop_back();
+        // Safe because:
+        //
+        // * Any slice passed to `drop_in_place` is valid; the second case has
+        //   `len <= front.len()` and returning on `len > self.len()` ensures
+        //   `begin <= back.len()` in the first case
+        // * The head of the VecDeque is moved before calling `drop_in_place`,
+        //   so no value is dropped twice if `drop_in_place` panics
+        unsafe {
+            if len > self.len() {
+                return;
+            }
```

Comment:
This is super minor but shouldn't this be outside the unsafe block? :sweat_smile:

I get keeping the below lines in one unsafe block for consistency, but conceptually, the idea is that no unsafe code is run if you don't actually need to truncate anything.

Reply:
Also… shouldn't this be `>=` instead of just `>`? I figure that it'll work either way, but still.

Reply:
Ah, I have no strong opinion on inside/outside for the check. Certainly there's nothing unsafe about it or anything before the `if/else` block, it was just inside to match the new `Vec::truncate`. I don't think it's necessarily unreasonable though, since you can't change the length check without violating the safety guarantees (as opposed to the check just avoiding unnecessary truncation).

The `>` vs `>=` is based on [this comment](https://github.com/rust-lang/rust/pull/64375#issuecomment-530525103); I haven't actually confirmed, but in theory this makes the `VecDeque::truncate(0)` case (used in clear) unconditional and thus easier to optimize. I'm happy to check that though, since it's definitely different enough that it might not hold.

Reply:
Ah, I didn't even consider the fact that the check is necessary to ensure safety. Makes sense to leave it in the block, then.

Reply:
In terms of the optimisation, I'll trust what's been done already.